src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -991,44 +991,103 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
       }
     }
 
+    /* Generate string concatenation code based on the strategy specified at `-YstringConcat` */
     def genStringConcat(tree: Tree): BType = {
       lineNumber(tree)
       liftStringConcat(tree) match {
-        // Optimization for expressions of the form "" + x.  We can avoid the StringBuilder.
+        // Optimization for expressions of the form "" + x
         case List(Literal(Constant("")), arg) =>
           genLoad(arg, ObjectRef)
           genCallMethod(String_valueOf, InvokeStyle.Static, arg.pos)
 
         case concatenations =>
-          val approxBuilderSize = concatenations.map {
-            case Literal(Constant(s: String)) => s.length
-            case Literal(c @ Constant(value)) if c.isNonUnitAnyVal => String.valueOf(c).length
-            case _ =>
-              // could add some guess based on types of primitive args.
-              // or, we could stringify all the args onto the stack, compute the exact size of
-              // the StringBuilder.
-              // or, just let https://openjdk.java.net/jeps/280 (or a re-implementation thereof in our 2.13.x stdlib) do all the hard work at link time
-              0
-          }.sum
-          bc.genStartConcat(tree.pos, approxBuilderSize)
-          def isEmptyString(t: Tree) = t match {
-            case Literal(Constant("")) => true
-            case _ => false
-          }
-          for (elem <- concatenations if !isEmptyString(elem)) {
-            val loadedElem = elem match {
+
+          val concatArguments = concatenations.view
+            .filterNot(_ == Literal(Constant("")))  // empty strings are no-ops in concatenation
+            .map {
               case Apply(boxOp, value :: Nil) if currentRun.runDefinitions.isBox(boxOp.symbol) =>
                 // Eliminate boxing of primitive values. Boxing is introduced by erasure because
                 // there's only a single synthetic `+` method "added" to the string class.
                 value
+              case other => other
+            }
+            .toList
+
+          if (currentSettings.YstringConcat.value == "inline" ||
+              currentSettings.YstringConcat.value == "detect" && currentSettings.target.value == "8") {
```

Comment:
Should this sort of thing be cached somewhere? Is there a good way to error and report to the user that they want something nonsensical if they have `-target:8 -YstringConcat:indy`?

Reply:
Is the flag really needed? Seems like the indy string concat should just always be used on target >= 9 and that's it.

Reply:
`javac` has a `stringConcat` compile option. If no one cares about having that, dropping the option is simpler. 😄

Reply:
I dropped the flag and now use `backendUtils.classfileVersion >= asm.Opcodes.V9` as the criteria for indy string concat.